experimental/transliteration/src/transliterator/mod.rs

```diff
@@ -0,0 +1,1417 @@
+        }
+
+        let mut matcher = matcher.finish_key();
+
+        if !self.post_matches(&mut matcher, &mut match_data, vt) {
+            return None;
+        }
+
+        Some((match_data, matcher))
+    }
+
+    /// Returns whether the ante context matches or not. Fills in `match_data` if applicable.
+    ///
+    /// This uses reverse matching.
+    fn ante_matches(
+        &self,
+        matcher: &mut impl Utf8Matcher<Reverse>,
+        match_data: &mut MatchData,
+        vt: &VarTable,
+    ) -> bool {
+        if self.ante.is_empty() {
+            // fast path for empty queries, which always match
+            return true;
+        }
+        helpers::rev_match_encoded_str(&self.ante, matcher, match_data, vt)
+    }
+
+    /// Returns whether the post context matches or not. Fills in `match_data` if applicable.
+    ///
+    /// This uses forward matching.
+    fn post_matches(
+        &self,
+        matcher: &mut impl Utf8Matcher<Forward>,
+        match_data: &mut MatchData,
+        vt: &VarTable,
+    ) -> bool {
+        if self.post.is_empty() {
+            // fast path for empty queries, which always match
+            return true;
+        }
+        helpers::match_encoded_str(&self.post, matcher, match_data, vt)
+    }
+
+    /// Returns whether the post context matches or not. Fills in `match_data` if applicable.
+    ///
+    /// This uses forward matching.
+    fn key_matches(
+        &self,
+        matcher: &mut impl Utf8Matcher<Forward>,
+        match_data: &mut MatchData,
+        vt: &VarTable,
+    ) -> bool {
+        if self.key.is_empty() {
+            // fast path for empty queries, which always match
+            return true;
+        }
+        helpers::match_encoded_str(&self.key, matcher, match_data, vt)
+    }
+}
+
+mod helpers {
+    use super::*;
+
+    /// Returns the index of the first encoded char in `s`, if there is one. Returns `None` if the
+    /// passed string is pure (contains no encoded special constructs).
+    pub(super) fn find_encoded(s: &str) -> Option<usize> {
+        // note: full-match (i.e., if this function returns Some(_) or None, could be
+        // precomputed + stored at datagen time (there could eg be a reserved char that is at the
+        // start/end of key <=> key is completely pure)
+        s.char_indices()
+            .find(|(_, c)| VarTable::ENCODE_RANGE.contains(c))
+            .map(|(i, _)| i)
+    }
+
+    /// Returns the index of the char to the right of the first (from the right) encoded char in
+    /// `s`, if there is one.
+    /// Returns `None` if the passed string is pure (contains no encoded special constructs).
+    pub(super) fn rev_find_encoded(s: &str) -> Option<usize> {
+        s.char_indices()
+            .rfind(|(_, c)| VarTable::ENCODE_RANGE.contains(c))
+            .map(|(i, c)| i + c.len_utf8())
+    }
+
+    /// Recursively estimates the size of the replacement string.
+    pub(super) fn estimate_replacement_size(
```

Comment:
Since the code path is already there, I'd like to keep this until there's better benchmarks. #3957 tracks this and is added as a comment directly above the entrypoint into `estimate_replacement_size`, so removing/keeping it can be compared easily. If you're strongly against it I could also just link this commit in that issue instead.